src/Strategy/Go/GoModGraph.hs

```diff
@@ -0,0 +1,189 @@
+import Data.Set (Set, fromList, member, notMember)
+import Data.String.Conversion (toText)
+import Data.Text (Text)
+import Data.Void (Void)
+import DepTypes (
+  DepType (GoType),
+  Dependency (..),
+  VerConstraint (CEq),
+ )
+import Effect.Exec (
+  AllowErr (Never),
+  Command (..),
+  Exec,
+  ExecErr (CommandParseError),
+  execParser,
+  execThrow,
+ )
+import Graphing qualified
+import Path
+import Strategy.Go.GoList (GoListModule (GoListModule, isIndirect, isMain, path, version))
+import Strategy.Go.Gomod (PackageVersion (..), parsePackageVersion)
+import Strategy.Go.Transitive (decodeMany)
+import Text.Megaparsec (
+  MonadParsec (eof),
+  Parsec,
+  empty,
+  optional,
+  parse,
+  sepEndBy,
+  some,
+  try,
+  (<|>),
+ )
+import Text.Megaparsec.Char (alphaNumChar, char)
+import Text.Megaparsec.Char.Lexer qualified as Lexer
+import Types (GraphBreadth (..))
+
+-- * Parsing
+
+type Parser = Parsec Void Text
+
+sc :: Parser ()
+sc = Lexer.space (void $ some $ char ' ') empty empty
+
+lexeme :: Parser a -> Parser a
+lexeme = Lexer.lexeme sc
+
+symbol :: Text -> Parser Text
+symbol = Lexer.symbol sc
+
+-- * Commands
+
+goModGraphCmd :: Command
+goModGraphCmd =
+  Command
+    { cmdName = "go"
+    , cmdArgs = ["mod", "graph"]
+    , cmdAllowErr = Never
+    }
+
+goListJsonCmd :: Command
+goListJsonCmd =
+  Command
+    { cmdName = "go"
+    , cmdArgs = ["list", "-m", "-json", "all"]
+    , cmdAllowErr = Never
+    }
+
+data GoGraphMod = GoGraphMod
+  { goGraphModName :: Text
+  , goGraphModVersion :: Maybe PackageVersion
+  }
+  deriving (Show, Eq, Ord)
+
+-- * Parsers
+
+-- | Parses go mode entry from 'go mod graph'.
+parseGoGraphMod :: Parser GoGraphMod
+parseGoGraphMod = do
+  name <- toText <$> lexeme (some (alphaNumChar <|> char '/' <|> char '-' <|> char '.' <|> char '_' <|> char '~'))
+  version <- optional $ do
+    _ <- try $ symbol "@"
+    lexeme (parsePackageVersion lexeme)
+  pure $ GoGraphMod name version
+
+toGoModVersion :: Text -> Maybe PackageVersion
+toGoModVersion modVersion = case parse (parsePackageVersion lexeme) "go module version" modVersion of
+  Left _ -> Nothing
+  Right vc -> Just vc
+
+-- | Parses output of 'go mod graph'.
+parseGoModGraph :: Parser [(GoGraphMod, GoGraphMod)]
+parseGoModGraph = parseGoModPair `sepEndBy` (symbol "\n" <|> symbol "\r") <* eof
+  where
+    parseGoModPair :: Parser (GoGraphMod, GoGraphMod)
+    parseGoModPair = (,) <$> parseGoGraphMod <*> parseGoGraphMod
+
+-- Builds graph from edges, main module, and selected module version set.
+buildGraph :: [(GoGraphMod, GoGraphMod)] -> Maybe GoGraphMod -> Set GoGraphMod -> Set GoGraphMod -> Bool -> Graphing.Graphing Dependency
+buildGraph fromToMods mainMod selectedMods directMods applyMVS =
+  Graphing.gmap toDependency
+    . Graphing.induceJust
+    . Graphing.gmap withSelection
+    . Graphing.promoteToDirect (`member` directMods)
+    . Graphing.shrink (\m -> m /= fromMaybe impossibleMod mainMod)
```

Comment:
We always expect to find a main module, right? If so, we should error out if we don't find one, rather than silently ignoring (i.e. pass `mainMod :: GoGraphMod` to `buildGraph` rather than `mainMod :: Maybe GoGraphMod`)